controllers/application_poller_controller.go

```diff
@@ -0,0 +1,278 @@
+	"github.com/thestormforge/optimize-controller/v2/internal/experiment"
+	"github.com/thestormforge/optimize-controller/v2/internal/scan"
+	"github.com/thestormforge/optimize-controller/v2/internal/server"
+	"github.com/thestormforge/optimize-controller/v2/internal/sfio"
+	"github.com/thestormforge/optimize-controller/v2/internal/version"
+	"github.com/thestormforge/optimize-go/pkg/api"
+	applications "github.com/thestormforge/optimize-go/pkg/api/applications/v2"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Poller struct {
+	client        client.Client
+	log           logr.Logger
+	apiClient     applications.API
+	kubectlExecFn func(cmd *exec.Cmd) ([]byte, error)
+}
+
+func NewPoller(kclient client.Client, logger logr.Logger) (*Poller, error) {
+	appAPI, err := server.NewApplicationAPI(context.Background(), version.GetInfo().String())
+	if err != nil {
+		logger.Info(err.Error())
+		return &Poller{log: logger}, nil
+	}
+
+	return &Poller{
+		client:    kclient,
+		apiClient: appAPI,
+		log:       logger,
+	}, nil
+}
+
+func (p *Poller) Start(ch <-chan struct{}) error {
+	if p.apiClient == nil {
+		<-ch
+		return nil
+	}
+	p.log.Info("Starting application poller")
+
+	ctx := context.Background()
+
+	query := applications.ActivityFeedQuery{}
+	query.SetType(applications.TagScan, applications.TagRun)
+	subscriber, err := p.apiClient.SubscribeActivity(ctx, query)
+	if err != nil {
+		p.log.Error(err, "unable to connect to application service")
+		return nil
+	}
+
+	activityCh := make(chan applications.ActivityItem)
+	go subscriber.Subscribe(ctx, activityCh)
+
+	for {
+		select {
+		case <-ch:
+			return nil
+		case activity := <-activityCh:
+			// We'll skip any janky activities that may come through
+			// TODO we should figure out the why for this
+			if activity.ID == "" {
+				continue
+			}
+
+			p.handleActivity(ctx, activity)
+		}
+	}
+}
+
+// handleActivity performs the task required for each activity.
+// When an ActivityItem is tagged with scan, the generation workflow is used to generate an experiment and the result
+// is converted into an api.Template consisting of parameters and metrics.
+// When an ActivityItem is tagged with run, the previous scanned template results are merged with
+// the results of an experiment generation workflow. Following this, the generated resources are applied/created
+// in the cluster.
+// note, rbac defined in cli/internal/commands/grant_permissions/generator
+func (p *Poller) handleActivity(ctx context.Context, activity applications.ActivityItem) {
+	// We always want to delete the activity after having received it
+	defer func() {
+		if err := p.apiClient.DeleteActivity(ctx, activity.URL); err != nil {
+			p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to delete activity", err), activity.URL)
```

Comment:
This should probably just log the error instead of trying to update the activity it just failed to delete.